File: ios/Sightread/Sightread/Services/GroqVisionService.swift
import Foundation

struct GroqVisionService: VisionAIService {
  let apiKey: String
  private let model = "meta-llama/llama-4-scout-17b-16e-instruct"

  func analyze(jpegData: Data, prompt: String) async throws -> String {
    guard !apiKey.isEmpty else { throw VisionAIError.missingAPIKey }
    let base64 = jpegData.base64EncodedString()
    guard let url = URL(string: "https://api.groq.com/openai/v1/chat/completions") else { throw VisionAIError.invalidResponse }
    let body: [String: Any] = [
      "model": model, "max_tokens": 300,
      "messages": [["role": "user", "content": [
        ["type": "text", "text": prompt],
        ["type": "image_url", "image_url": ["url": "data:image/jpeg;base64,\(base64)"]],
      ]]],
    ]
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
    request.httpBody = try JSONSerialization.data(withJSONObject: body)
    let (data, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
      let code = (response as? HTTPURLResponse)?.statusCode ?? -1
      throw VisionAIError.httpError(code, String(data: data, encoding: .utf8) ?? "")
    }
    guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
          let choices = json["choices"] as? [[String: Any]],
          let message = choices.first?["message"] as? [String: Any],
          let text = message["content"] as? String, !text.isEmpty else { throw VisionAIError.invalidResponse }
    return text.trimmingCharacters(in: .whitespacesAndNewlines)
  }
}

File: android/Sightread/app/src/main/java/com/meta/wearable/dat/externalsampleapps/sightread/ai/GroqVisionService.kt
package com.meta.wearable.dat.externalsampleapps.sightread.ai

import android.util.Base64
import com.meta.wearable.dat.externalsampleapps.sightread.network.ApiClient
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

class GroqVisionService(private val apiKey: String) : VisionAIService {
  override suspend fun analyze(jpegData: ByteArray, prompt: String): String =
      withContext(Dispatchers.IO) {
        require(apiKey.isNotBlank()) { "Add a Groq API key in Settings." }
        val base64 = Base64.encodeToString(jpegData, Base64.NO_WRAP)
        val content =
            JSONArray()
                .put(JSONObject().put("type", "text").put("text", prompt))
                .put(
                    JSONObject()
                        .put("type", "image_url")
                        .put(
                            "image_url",
                            JSONObject().put("url", "data:image/jpeg;base64,$base64"),
                        ),
                )
        val body =
            JSONObject()
                .put("model", "meta-llama/llama-4-scout-17b-16e-instruct")
                .put("max_tokens", 300)
                .put(
                    "messages",
                    JSONArray().put(JSONObject().put("role", "user").put("content", content)),
                )
                .toString()
        val result =
            ApiClient.postJson(
                "https://api.groq.com/openai/v1/chat/completions",
                body,
                mapOf("Authorization" to "Bearer $apiKey"),
            )
        if (!result.isSuccess) error("API error ${result.code}: ${result.body}")
        JSONObject(result.body)
            .getJSONArray("choices")
            .getJSONObject(0)
            .getJSONObject("message")
            .getString("content")
            .trim()
      }
}
